packages/core/ios/RNEmbrace/SpanRepository.swift

```diff
@@ -0,0 +1,90 @@
+import Foundation
+import OSLog
+import OpenTelemetryApi
+
+// Should not get hit under normal circumstances, add as a guard against misinstrumentation
+private let MAX_STORED_SPANS = 10000
+
+class SpanRepository {
+    private let activeSpansQueue = DispatchQueue(label: "io.embrace.RNEmbrace.activeSpans",
+                                                 attributes: .concurrent)
+    private let completedSpansQueue = DispatchQueue(label: "io.embrace.RNEmbrace.completedSpans",
+                                                    attributes: .concurrent)
+    private var activeSpans = [String: Span]()
+    private var completedSpans = [String: Span]()
+    private var log = OSLog(subsystem: "Embrace", category: "SpanRepository")
+
+    init() {
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(onSessionEnded),
+            // TODO possible to reference this constant? https://github.com/embrace-io/embrace-apple-sdk/blob/main/Sources/EmbraceCore/Session/SessionController.swift#L13
+            name: Notification.Name("embrace.session.will_end"),
+            object: nil
+        )
+    }
+
+    deinit {
+        NotificationCenter.default.removeObserver(self)
+    }
+
+    private func getKey(_ span: Span) -> String {
+        return "\(span.context.spanId.hexString)_\(span.context.traceId.hexString)"
+    }
+
+    func get(spanId: String) -> Span? {
+        var span: Span?
+
+        activeSpansQueue.sync {
+          span = activeSpans[spanId]
+        }
+
+        if span == nil {
+            completedSpansQueue.sync {
+                span = completedSpans[spanId]
+            }
+        }
+
+        return span
+    }
+
+    func spanStarted(span: Span) -> String {
+        let key = getKey(span)
+
+        // TODO does concurrent access matter for grabbing count?
```

Comment:
No it doesn't at all. Even though having the actual exact number is ok, we apply limits to OTel stuff mostly to prevent any problems on the SDKs (like growing memory too much, or using too much disk space). Actual limits (if we'd charge by X amount of spans/events/etc per session) should be also covered in the backend.
FWIW, on the SDK side, we have an in-memory counter so we don't go constantly to the database for this validation. And even though you can do this in a "thread-safe" manner (having a sync queue / lock accessing and setting the actual number of `activeSpans.count` to have the exact number) I believe this is not necessary.

Reply:
sounds good, I'll leave as is then